### ISSUE TITLE
Some problems when collecting patterns

### ISSUE THREAD

**Issue by yangkaiarvin** [crigroup/ensenso#19]
I created a subscriber to listen to the topic "/camera/pattern/stereo", but there is something wrong when it  comes to` last_stereo_pattern_ = get_pattern_buffers.result()[itmStereo][0].asJson (true); `in ensenso_grabber.cpp with the camera seeing the calibration pattern. The program stopped with no response. How can I get the pattern data from "/camera/pattern/stereo"?

Best Wishes!

**Comment by fsuarez6** [comment]
Yes, this is not working as expected. Could you please share the details of your system? In particular the following:

|  | Version |
| ---- | --------- |
| Camera | N35-805-16-BL |
| NxLib | 2.0.147 |
| uEyeDriver | 4.81.18 |
| CodeMeter | 6.40.2402 |

You can get these details running:
```
$ nxView
```
and hovering the mouse over the lower-right corner

**Comment by yangkaiarvin** [comment]
Version
Camera N30-602-16-BL
NxLib 2.0.147
uEyeDriver  4.82.16
CodeMeter 6.40.2402

When we use the `collect_srv_ = nh_.advertiseService("collect_pattern", &EnsensoDriver::collectPatternCB, this);`, sometimes we will have `getPatternCount()>0` but mostly the patterncount was 0. And the program keep running with no response.

**Comment by fsuarez6** [comment]
We have two use cases for the `/camera/collect_pattern` service:

## Camera-robot calibration or handeye calibration
You want determine the position of the camera relative to a robot. The camera can either be fixed with respect to the robot's base or it can be mounted on the robot's hand and move with it.
More details: 
- [How to: Handeye calibration](https://www.ensenso.com/manual/howto_handeyecalibration.htm)
- [Pattern buffer](https://www.ensenso.com/manual/_patternbuffer.htm)

### MWE for the calibration
Start the ensenso driver:
```bash
$ roslaunch ensenso viewer.launch rqt:=False rviz:=False
```
Configure the driver to only stream the images and switch off the projector. The front light will depend of your illumination conditions:
```bash
$ rosrun dynamic_reconfigure dynparam set /camera/ensenso_driver "{'Images':true, 'Cloud':false, 'Projector':false, 'FrontLight':false}"
```
The `/camera/left/image_raw` topic contains the raw image and when the pattern has been detected an overlay will appear in the image:
```bash
rosrun image_view image_view image:=/camera/left/image_raw
```
![image_raw_screenshot](https://user-images.githubusercontent.com/4430326/35952676-20438d2a-0cbc-11e8-8e6a-579114ca1ebe.png)

Collect the pattern observations. For the first call, **clear the buffer**:
```bash
$ rosservice call /camera/collect_pattern "add_to_buffer: true
clear_buffer: true
decode: true
grid_spacing: 0.0"
# Output
success: True
pattern_count: 1
grid_spacing: 18.75
```
For the subsequent calls, **do not** clear the buffer:
```bash
$ rosservice call /camera/collect_pattern "add_to_buffer: true
clear_buffer: false
decode: true
grid_spacing: 0.0"
# Output
success: True
pattern_count: 2
grid_spacing: 18.75
```
The `pattern_count` should increase with every observation.
For each observation you must know the **robot pose**.
Once you have enough observations (we usually collect 25) you can call the `/camera/calibrate_handeye` service.

## Pattern pose estimation
You just need the pattern pose with respect to the camera frame.

### MWE for the pattern pose estimation
Start the ensenso driver, configure it to only stream the images and switch off the projector.

Collect the pattern:
```bash
$ rosservice call /camera/collect_pattern "add_to_buffer: true
clear_buffer: true
decode: true
grid_spacing: 0.0"
# Output
success: True
pattern_count: 1
grid_spacing: 18.75
```
Estimate the pattern pose:
```bash
$ rosservice call /camera/estimate_pattern_pose "average: false" 
# Output
success: True
pose: 
  position: 
    x: 0.140828834732
    y: 0.154744567362
    z: 1.01843195859
  orientation: 
    x: -0.158285623859
    y: -0.183593665657
    z: -0.708303946633
    w: 0.662981558111
```

Hope this helps!

**Comment by yangkaiarvin** [comment]
Problems solved, thank you!
